src/levanter/models/llama.py

```diff
@@ -0,0 +1,405 @@
+    def __call__(self, x: NamedArray, mask: Optional[NamedArray], position_ids, *args):
+        q = self.q_proj(x)  # TODO: rearrange and possibly rename
+        k = self.k_proj(x)
+        v = self.v_proj(x)
+
+        cos, sin = self.rotary_emb(v, seq_len=self.config.seq_len)
+
+        q, k = _apply_rotary_pos_emb(q, k, cos, sin, position_ids)
+
+        scale = jax.lax.rsqrt(float(self.config.HeadSize.size))
+
+        # do this first to help keep FP values small
+        q = q * scale
+
+        q = q.astype(jnp.float32)
+        k = k.astype(jnp.float32)
+        k = k.rename({"position": "key_position"})
+
+        attn_scores = hax.dot("head_size", q, k)
+
+        if mask is not None:
+            attn_scores = attn_scores + (1.0 - mask) * -1e9
+
+        attn_scores = attn_scores.astype(jnp.float32)
+        attn_weights = hnn.softmax(attn_scores, axis="key_position").astype(x.dtype)
+        # There's no dropout in llama attention, compared with Gpt2 attention
+
+        attn_output = hax.dot("key_position", attn_weights, v)
+
+        attn_output = self.o_proj(attn_output)
+        return attn_output
+
+
+class LlamaDecoderLayer(StateDictSerializationMixin, eqx.Module):
+    config: LlamaConfig = eqx.static_field()
+    attn: LlamaAttention
+    mlp: LlamaMlp
+    ln_1: hnn.LayerNorm  # input layernorm
+    ln_2: hnn.LayerNorm  # post attention layernorm
+
+    @staticmethod
+    def init(config: LlamaConfig, *, key) -> "LlamaDecoderLayer":
+        k_attn, k_mlp = jrandom.split(key, 2)
+
+        attn = LlamaAttention.init(config, key=k_attn)
+        mlp = LlamaMlp.init(config.Embed, config.Mlp, config.activation_function, key=k_mlp, use_bias=config.use_bias)
+        ln_1 = hnn.LayerNorm.init(config.Embed, eps=config.layer_norm_epsilon, use_bias=config.use_bias)
+        ln_2 = hnn.LayerNorm.init(config.Embed, eps=config.layer_norm_epsilon, use_bias=config.use_bias)
+
+        return LlamaDecoderLayer(config, attn, mlp, ln_1, ln_2)
+
+    @named_call
+    def __call__(self, x: NamedArray, mask: Optional[NamedArray], position_ids, *args):
+        residual = x
+        x = self.ln_1(x)
+
+        # self attention and skip connection
+        attn_output = self.attn(x=x, mask=mask, position_ids=position_ids)
+        x = residual + attn_output
+
+        # MLP and skip connection
+        residual = x
+        x = self.ln_2(x)
+        mlp_output = self.mlp(x)
+        output = residual + mlp_output
+        return output
+
+
+class LlamaTransformer(StateDictSerializationMixin, eqx.Module):
+    config: LlamaConfig = eqx.static_field()
+    layers: Stacked[LlamaDecoderLayer]
+    ln_f: hnn.LayerNorm
+
+    @staticmethod
+    def init(config: LlamaConfig, *, key) -> "LlamaTransformer":
+        # TODO: here it reports an error that is related to _get_rotary_emb() in LlamaAttention
+        # TypeError: Output from batched function Axis(name='head_size', size=4) with type <class 'haliax.axis.Axis'> is not a valid JAX type
```

Comment:
`Stacked.init` reports an error that is related to _get_rotary_emb() in LlamaAttention. Here's the full stacktrace:

```
def test_llama_transformer():
        llama_config = _get_llama_config()
        key = random.PRNGKey(0)
>       llama_transformer = LlamaTransformer.init(config=llama_config, key=key)

tests/test_llama.py:156: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/levanter/models/llama.py:275: in init
    layers = Stacked.init(config.Layers, LlamaDecoderLayer, gradient_checkpointing=config.gradient_checkpointing)(
/nlp/scr/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/haliax/nn/scan.py:69: in fn
    stacked = haliax.vmap(module.init, Block)(*args, **kwargs)
/nlp/scr/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/haliax/hof.py:270: in wrapped_vmap_fn
    result = jax.vmap(
/nlp/scr/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/_src/api.py:1240: in vmap_f
    out_flat = batching.batch(
/nlp/scr/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/_src/linear_util.py:203: in call_wrapped
    ans = gen.send(ans)
/nlp/scr/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/_src/interpreters/batching.py:600: in _batch_inner
    out_vals = map(partial(from_elt, trace, axis_size), outs, out_dim_dests)
/nlp/scr/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/jax/_src/interpreters/batching.py:237: in from_elt
    return matchaxis(trace.axis_name, axis_size, x_.batch_dim, spec, x_.val)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

axis_name = <object object at 0x7f22a3423e10>, sz = 32, src = None, dst = 0, x = Axis(name='head_size', size=4), sum_match = False

    def matchaxis(axis_name, sz, src, dst, x, sum_match=False):
      if dst == pile_axis:
        x = bdim_at_front(x, src, sz)
        elt_ty = x.aval.update(shape=x.shape[1:])
        aval = PileTy(core.Var(0, '', core.ShapedArray((), np.dtype('int32'))),
                      x.shape[0], elt_ty)
        return Pile(aval, x)
      try:
        _ = core.get_aval(x)
      except TypeError as e:
>       raise TypeError(f"Output from batched function {repr(x)} with type "
                        f"{type(x)} is not a valid JAX type") from e
E       jax._src.traceback_util.UnfilteredStackTrace: TypeError: Output from batched function Axis(name='head_size', size=4) with type <class 'haliax.axis.Axis'> is not a valid JAX type
E       
E       The stack trace below excludes JAX-internal frames.
E       The preceding is the original exception that occurred, unmodified.
```

Reply:
this is a new one for me. is this changed pushed?

Reply:
pretty sure it's this (or perhaps similar problems):

https://github.com/stanford-crfm/levanter/blob/6e222b7297250126cc17a8ab5c834c12ded592bf/src/levanter/models/llama.py#L97C3-L98C14

those need to be static fields

Reply:
Will make a try